File: pull-model.py
import subprocess
import sys
import shutil
import re

MODEL = sys.argv[1] if len(sys.argv) > 1 else "llama3"

def show_progress(line: str):
    m = re.search(r"(\d+)%.*?(\d+\.?\d*)\s*(GB|MB|KB)/s.*?(\d+)s", line)
    if m:
        pct, speed_val, unit, eta = m.group(1), m.group(2), m.group(3), m.group(4)
        speed = f"{speed_val} {unit}/s"
        bar_width = 30
        filled = int(pct) * bar_width // 100
        bar = "\u2588" * filled + " " * (bar_width - filled)
        print(f"\r  Pulling... {pct}% \u2595{bar}\u258f {speed}  ETA: {eta}s  ", end="", flush=True)
    elif any(kw in line for kw in ("pulling manifest", "verifying", "writing manifest", "success")):
        print(f"\r{line}")
    elif line.strip():
        print(f"\r{line}")

def main():
    print("\n  Ollama Model Downloader\n")
    print(f"  Model: {MODEL}\n")

    if not shutil.which("ollama"):
        print("  [!] Ollama not found. Install from https://ollama.ai")
        input("\n  Press Enter to exit...")
        sys.exit(1)

    confirm = input(f"  Pull model '{MODEL}'? (Y/n): ").strip().lower()
    if confirm == "n":
        return

    print()
    proc = subprocess.Popen(
        ["ollama", "pull", MODEL],
        stdout=subprocess.PIPE,
        stderr=subprocess.STDOUT,
        text=True,
        bufsize=1,
    )
    for line in proc.stdout or []:
        show_progress(line)
    proc.wait()

    if proc.returncode == 0:
        print("\n  \u2713 Download complete")
    else:
        print("\n  [!] Download failed")

    input("\n  Press Enter to exit...")

if __name__ == "__main__":
    main()

File: pull-model.js
const { execSync, spawn } = require("child_process");
const readline = require("readline");

const MODEL = process.argv[2] || "llama3";

function confirm(query) {
  const rl = readline.createInterface({ input: process.stdin, output: process.stdout });
  return new Promise((resolve) => rl.question(query, (a) => { rl.close(); resolve(a.trim().toLowerCase() !== "n"); }));
}

function showProgress(line) {
  const m = line.match(/(\d+)%.*?(\d+\.?\d*)\s*(GB|MB|KB)\/s.*?(\d+)s/);
  if (m) {
    const pct = m[1], speed = `${m[2]} ${m[3]}/s`, eta = m[4];
    const barWidth = 30;
    const filled = (parseInt(pct) * barWidth / 100) | 0;
    const bar = "\u2588".repeat(filled) + " ".repeat(barWidth - filled);
    process.stdout.write(`\r  Pulling... ${pct}% \u2595${bar}\u258f ${speed}  ETA: ${eta}s  `);
  } else if (/pulling manifest|verifying|writing manifest|success/.test(line)) {
    console.log(`\r${line}`);
  } else if (line.trim()) {
    console.log(`\r${line}`);
  }
}

async function main() {
  console.log(`\n  Ollama Model Downloader\n  Model: ${MODEL}\n`);

  try { execSync("where ollama", { stdio: "ignore" }); }
  catch {
    console.log("  [!] Ollama not found. Install from https://ollama.ai");
    return;
  }

  if (!(await confirm(`  Pull model '${MODEL}'? (Y/n): `))) return;

  console.log();
  const proc = spawn("ollama", ["pull", MODEL], { stdio: ["ignore", "pipe", "pipe"], shell: true });

  proc.stdout.on("data", (d) => d.toString().split("\n").forEach(showProgress));
  proc.stderr.on("data", (d) => d.toString().split("\n").forEach(showProgress));

  proc.on("exit", (code) => {
    console.log(code === 0 ? "\n  \u2713 Download complete" : "\n  [!] Download failed");
  });
}

main();
